### PR TITLE
Create deprecating-components.md

### PR DESCRIPTION
**Pull request by broccolinisoup**
Closes https://github.com/github/primer/issues/1593

This PR is based on deprecating React components [ADR](https://github.com/primer/react/blob/main/contributor-docs/adrs/adr-006-drafts.md) , my notes from chatting with PRC folks about the process as well as my first time deprecating a component experience. So this is a pretty much raw document that I would love to get your thoughts on 🙏🏼 

Note: I copy-pasted quite a few info from the [ADR](https://github.com/primer/react/blob/main/contributor-docs/adrs/adr-006-drafts.md), thanks to @siddharthkp :) 

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [inline contributor-docs/deprecating-components.md:9]:
```suggestion
import { ActionList } from "@primer/react/drafts"
```

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [inline contributor-docs/deprecating-components.md:61]: @lesliecdubs I pretty much copied your message from [the slack message you posted](https://github.slack.com/archives/C03M1LZ0UR3/p1670858880139779) and I am not sure if we would like to add extra information here. Could I please get your thoughts on that too? Thank you!

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [inline contributor-docs/deprecating-components.md:59]: Do we provide any other support for teams except proving these codemods? Do they actually go and update the usages themselves or do we do it? 

Also, is there any other practise we do or would like to introduce to encourage teams switching using the new components?

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [inline contributor-docs/deprecating-components.md:1]: I imagine there could also be a second reason for deprecating a component: when a component is no longer recommended for use (even in cases where we may not offer a direct replacement for it). Perhaps we could broaden this explanation to something like:

```suggestion
Primer components may occasionally be deprecated when they are no longer recommended for use. In some cases, we may recommend replacing a deprecated component with a new one. New replacement components are typically built in a parallel bundle to avoid breaking existing components, and are promoted to the main bundle when it's time for migration.
```

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [inline contributor-docs/deprecating-components.md:3]:
```suggestion
## Developing a replacement component
```

We're planning to consolidate the "draft" and "experimental" lifecycle stages into one stage (https://github.com/github/primer/issues/1463), so I'm suggesting this change for clarity.

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:5]: @siddharthkp instead of recommending these components go into `src/drafts`, do you think we should be encouraging folks to build directly in the new experimental components repo instead? https://github.com/github/experimental-react-components

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [inline contributor-docs/deprecating-components.md:16]:
```suggestion
When Primer maintainers are ready to deprecate a component, they will plan for an upcoming major release that will include the change. This includes preparing code changes and communications about the change.
```

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [inline contributor-docs/deprecating-components.md:18]: Do we need to add documentation for what to do in a case where we aren't replacing a deprecated component with a new one as well?

We could also add to this doc later if needed.

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:55]: Thanks for including this!

```suggestion
When we know if the newly developed component is ready to go in the main bundle and the old component is good to be deprecates in the next major release, we announce the deprecation and the promotion in the [Primer changelog](https://github.com/github/primer/discussions/categories/primer-changelog) (GitHub staff only) under the `Coming soon` section.

To ensure the impending deprecation is included in the Primer changelog, Primer maintainers should highlight any upcoming deprecations in their [weekly status updates](https://github.com/github/design-infrastructure/blob/main/how-we-work/planning-and-tracking-work/updates.md#weekly-status-updates-required) (GitHub staff only).
```

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [inline contributor-docs/deprecating-components.md:61]:
```suggestion
### Primer maintainers to alert Accessibility team about deprecations
```

Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:61]: For sure! Left a few suggestions.

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [inline contributor-docs/deprecating-components.md:68]:
```suggestion
As GitHub teams onboard to the [accessibility scorecard](https://github.com/github/engineering/discussions/2443) (GitHub staff only), we have a new opportunity to encourage teams to remove deprecated components and adopt new, more accessible ones.

Primer maintainers should reach out to #accessibility any time a component is planned for deprecation, or if there is a new component that we want to enforce usage around.
```

**Review by lesliecdubs — CHANGES_REQUESTED** [review]
Thank you so much for writing this up!

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:5]: Oh super keen to hear about this 👀

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:18]: Hmm that is great point! I wonder if we experienced such thing in Primer before? 🤔 What would be the alternative we can recommend here if we are not able to replace a component?

Any of the below at all?
- Temporary solutions where applicable
- Guideline on how to "replicate" the component's UI and functionality without making it a Primer component
- Recommending opening PRs for teams to fix the issues that made the component not usable

**Comment by broccolinisoup** [comment]
> Thank you so much for writing this up!

Thank you for a great review and suggestions @lesliecdubs! 🙏🏼  When I first created [the issue](https://github.com/github/primer/issues/1593,) I imagined that we could write this doc both for PRC and PVC but when I started working on deprecating UnderlineNav, I realised the lack of technical documentation of the process too. It is figurable though what to do when look into previously deprecated components but still there were quite a few things I missed and realised when the CI was failed. I just though it would be helpful to have a checklist to save some time really - instead of CI-driven development 😄 However, this way the doc became a PRC deprecation docs rather than Primer.

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:18]: I checked through [the deprecated components in our docs](https://primer.style/react/deprecated/ActionList), and it appears that in all cases we provided an alternative component to use even if it wasn't a one-to-one update (ex: for [deprecated `ChoiceFieldset`](https://primer.style/react/deprecated/ChoiceFieldset), we recommended to use totally different components: `CheckboxGroup` or `RadioGroup`).

Perhaps we could broaden the documentation a little bit, to cover two key cases?
1. when there is a draft/experimental component replacing a component we're deprecating - this is the case already covered by these docs
2. when there is a component being deprecated that doesn't have a one-to-one replacement, but where we recommend the use of another, different component in Primer instead

**Comment by lesliecdubs** [comment]
> However, this way the doc became a PRC deprecation docs rather than Primer.

Totally fair. Having this doc for only PRC is still better than nothing, and adding it here in `primer/react` makes it clear what these docs refer to  ✨

File: contributor-docs/deprecating-components.md
Comment (siddharthkp) [reply contributor-docs/deprecating-components.md:5]: Yes! 

But given the change in direction for experimental-components, i'm not sure if it should be `github/experimental-react-components` or our own `primer/react-experimental`, another ADR? 😅

Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:5]: Yes, probably. @justinbyo is going to be helping us figure out ownership models for shared React components looking forward.

In the meantime, are we comfortable shipping these docs with the direction to build direct in `experimental-react-components` for now, and we can always update if/when we make a change? We don't seem to deprecate components super often so I'm thinking that changing this guidance later wouldn't be a huge problem.

Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:5]: If we are going to update here to build the components in `experiment-react-components`, some items (the ones related to the new component) in the [checklist](https://github.com/primer/react/pull/2733/files#diff-1760a697b5e00b53e2aecf5aefbec00b6fd0a8fbee6a228fa3e74adfd33ed458R22) won't be valid anymore and need to be updated to make sure that the "Experimental React components" are becoming "Primer" components. (just as an example we have our custom unit tests and storybook format and with the new docs update a json file that has metadata of the components). I am just not sure if deprecating component docs is the best place to write about "Primer" components. I  _think_ it deserves its own documentation -  like "what are Primer" component standards - and we can link here?

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:18]: Sounds great! I wouldn't think of recommending a different component but that makes a lot of sense! 

How about this? We could add maybe just before the `Code side of the deprecation`?

> If we are not providing a one-to-one replacement for the component we are deprecating, we will be recommending the use of another, different component in Primer instead. We will make sure to communicate this in [Primer changelog](https://github.com/github/primer/discussions/categories/primer-changelog) (GitHub staff only) and deprecated component docs as well.

Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:18]: Messaging looks good! Suggesting a few tweaks to reduce the [first-person perspective](https://www.grammarly.com/blog/first-second-and-third-person/):

> If there is not a one-to-one replacement for the newly deprecated component, Primer maintainers will recommend the use of another, different component in Primer instead. This will be clearly communicate in the [Primer changelog](https://github.com/github/primer/discussions/categories/primer-changelog) (GitHub staff only) and deprecated component docs.

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:5]: Good point @broccolinisoup. In most cases we should have similar standards for components (including unit tests and Storybook) in the `experimental-react-components` repo as we do in `primer/react`. I'm admittedly not as familiar with all of the items in the checklist and whether they would apply exactly the same in `experimental-react-components`; what do you think @siddharthkp?

Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:5]: > In most cases we should have similar standards for components (including unit tests and Storybook) in the experimental-react-components repo as we do in primer/react

I am a fan of this! 

I wasn't sure how much we can "influence" `experimental-react-components` to follow our Primer component standards. 
From my observation, `experimental-react-components` differ from PRC in the way of [documenting components](https://github.com/github/experimental-react-components/blob/main/docs/documenting-components.md) and they don't include some [util unit test functions](https://github.com/primer/react/blob/main/src/utils/testing.tsx) as PRC do.   

Super keen to hear @siddharthkp's thoughts.

File: contributor-docs/deprecating-components.md
Comment (joshblack) [inline contributor-docs/deprecating-components.md:76]:
```suggestion
- Does the deprecated component accept new feature requests?
```

File: contributor-docs/deprecating-components.md
Comment (joshblack) [inline contributor-docs/deprecating-components.md:77]: Might be helpful to highlight that, once deprecated, the component is likely in a "maintenance mode" and will only receive bug fixes. Maybe something like:

```suggestion
  - No, once a component is deprecated the only changes it will receive are bug fixes.
```

**Review by joshblack — APPROVED** [review]
Looks great! I feel like this documents my understanding of the current deprecation process 👍 Just left a grammar suggestion and a note on the FAQ 👀

File: contributor-docs/deprecating-components.md
Comment (siddharthkp) [reply contributor-docs/deprecating-components.md:5]: > I wasn't sure how much we can "influence" experimental-react-components to follow our Primer component standards.

Oooh, we definitely should influence it a lot. We can even assist (and enforce) practices by baking them into the [react toolchain](https://github.com/github/experimental-react-components/tree/main/packages/%40github/react-toolchain).

However, we also keep the barrier for contribution pretty low. Experimental components are meant to be [pre-alpha in the lifecycle](https://primer.style/contribute/component-lifecycle/), more context in [ADR 008](https://github.com/primer/react/blob/main/contributor-docs/adrs/adr-008-experimental-components.md)

> From my observation, experimental-react-components differ from PRC in the way of [documenting components](https://github.com/github/experimental-react-components/blob/main/docs/documenting-components.md) and they don't include some [util unit test functions](https://github.com/primer/react/blob/main/src/utils/testing.tsx) as PRC do.

Some of this, like the docs, is intentional, we chose storybook docs over doctocat to make it easier (and less elaborate) to document than primer. The lack of test functions isn't part of the strategy, that's just because we've only had one iteration of this till now :)

Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:5]: >Oooh, we definitely should influence it a lot. We can even assist (and enforce) practices by baking them into the [react toolchain](https://github.com/github/experimental-react-components/tree/main/packages/%40github/react-toolchain).

Awesome to hear! 

>However, we also keep the barrier for contribution pretty low. Experimental components are meant to be [pre-alpha in the lifecycle](https://primer.style/contribute/component-lifecycle/), more context in [ADR 008](https://github.com/primer/react/blob/main/contributor-docs/adrs/adr-008-experimental-components.md)

This is a pretty good point!

>Some of this, like the docs, is intentional, we chose storybook docs over doctocat to make it easier (and less elaborate) to document than primer. The lack of test functions isn't part of the strategy, that's just because we've only had one iteration of this till now :)

Gotcha! 

It feels to me that moving experimental components from `react-experimental-components` is a slightly different process than deprecating components, especially around the practises.

Another question - if an existing component was to be re-written by a Primer engineer (following all Primer standards), would we still start in the `experimental-react-components` repo or in our own repos?

cc @lesliecdubs

File: contributor-docs/deprecating-components.md
Comment (lesliecdubs) [reply contributor-docs/deprecating-components.md:77]: What do y'all think: should this be _will receive_ or _may receive_? Not sure how dedicated we've been to handling all bugfixes in deprecated components, or just major functional bugs 🤔

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md]: Hello! @siddharthkp and @lesliecdubs

Seems like we still have questions regarding to `experimental-react-components` repo's practises and how they will make their way to Primer. Would it be okay to go ahead with this section as is (creating the components in the `drafts` folder) until we have a clear practise for `experimental-react-components`?  At least we will treat this document as the current practise and we can update as we go?

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md]: I think "may receive" sounds good to me. I like that it removes the certainty.
 
I updated to be "may receive" but please let me know if anyone has a concern 🙌🏻

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [inline contributor-docs/deprecating-components.md:61]: I updated this section to reflect @rezrah's [comments on versioning docs](https://github.com/primer/react/pull/2753/files#r1085554252) ✨

File: contributor-docs/deprecating-components.md
Comment (broccolinisoup) [inline contributor-docs/deprecating-components.md:1]: I am not sure if "Promoting" a good word for describing the process to move the deprecated component into the deprecated bundle. I thought it would be good to separate the process of deprecation (as in adding a `@deprecated` annotation) and moving the component into the deprecated folder.

Comment (broccolinisoup) [reply contributor-docs/deprecating-components.md:1]: cc @colebemis

**Comment by broccolinisoup** [comment]
I'd like to move forward with this PR if that is okay. This reflects the current process - We will need to come back and update the documentation around creating the components under the `drafts` folder when we have our unstable bundle is renamed to `experimental` from `drafts`. Or If we come up with an alternative solution we will need to update this doc. However currently, it all reflects the current process.

@lesliecdubs Since you requested changes before, it needs your approval  😇  Thank you!

**Review by lesliecdubs — APPROVED** [review]

**Pull request merged**
